fms_extras/utils/generation.py

```diff
@@ -0,0 +1,155 @@
+    cudagraphs: bool = False,
+) -> Tuple[torch.Tensor, int, float, float]:
+    """
+    A trivial generate function that can be used for validation/testing generation using paged attention
+
+    Args:
+        model: Callable or nn.Module
+            A function or nn.Module that takes a batch of input_ids and returns logits
+        input_ids_list: List[torch.Tensor]
+            A list of tensors, each tensor being for a single prompt
+        kv_cache_manager: PagedKVCacheManager
+            the paged KVCacheManager that handles management of the kv-cache for paged attention
+        max_seq_len: int
+            the max sequence length of the model
+        max_new_tokens: int
+            max tokens to generate
+        temperature: float
+            temperature of softmax when sampling
+        top_k: int
+            only search among top k tokens
+        do_sample: bool
+            multinomial sampling. False for greedy.
+        decode_model: Callable or nn.Module, optional
+            a model to used specifically for decode step. If not given, the model input will be used for decode step
+        cudagraphs: bool
+            if True, model is using cudagraphs and input will be padded, otherwise no cudagraphs is used and input is
+            not padded
+
+    Returns:
+    Tuple[torch.Tensor, int, float, float]
+        the resulting output tokens, the number of new tokens generated, the time to first token in seconds and the
+        total decode time
+    """
+    start_time = time.time()
+    if decode_model is None:
+        decode_model = model
+
+    bsize = len(input_ids_list)
+
+    if cudagraphs and bsize != 1:
+        raise NotImplementedError(
+            "cudagraphs is not yet supported for batch sizes greater than 1"
+        )
+
+    # Build padded batched input tensor
+    max_len = max([seq.size(0) for seq in input_ids_list])
+    n_pads_init = [max_len - seq.size(0) for seq in input_ids_list]
+    input_ids = torch.stack(
+        [F.pad(input_ids_list[i], (n_pads_init[i], 0)) for i in range(bsize)]
+    )
+
+    result = input_ids
+    next_input = input_ids
+    kwargs: MutableMapping[str, Any] = dict()
+    kwargs["cache_data"] = None
+    kwargs["use_cache"] = True
+    sequence_ids: Optional[List[int]] = None
+    block_mapping_max = ((max_len + max_new_tokens) // 16) + 1
+    for i in range(max_new_tokens):
+        if i == 1:
+            start_time = time.time()
+
+        input_ids = next_input[:, -max_seq_len:]
+
+        # compute the mask
+        if i == 0:
+            is_pad = input_ids == 0
+            mask = is_pad.unsqueeze(-1) == is_pad.unsqueeze(-2)
+            kwargs["mask"] = mask.tril(diagonal=0)
+        else:
+            kwargs["mask"] = None
+
+        # get the cache data and position ids if using cache
+        if sequence_ids is None:
+            num_tokens_per_sequence = torch.count_nonzero(input_ids.T, dim=0).tolist()
```

Comment:
i still don't love using non-zero to find non-pad tokens. A prompt could contain a 0, which means 'unk' in the llama tokenizer, and could mean something else with some other tokenizer.

Reply:
Because we take in the sequence (not padded), I can extract the num_tokens_per_sequence pretty easily just using the list of input_ids directly. For the mask, I can create it based on this `num_tokens_per_sequence`. I think this should simplify things, and this way we make no assumptions about 0s in the inputs.

Reply:
Fixed in latest commit, no assumptions made about token=0